driver-core/src/main/com/mongodb/internal/connection/SingleServerCluster.java

```diff
@@ -82,19 +82,21 @@ public void close() {
     private class DefaultServerStateListener implements ServerListener {
         @Override
         public void serverDescriptionChanged(final ServerDescriptionChangedEvent event) {
-            ServerDescription descriptionToPublish = event.getNewDescription();
-            if (event.getNewDescription().isOk()) {
+            ServerDescription newDescription = event.getNewDescription();
+            if (newDescription.isOk()) {
                 if (getSettings().getRequiredClusterType() != ClusterType.UNKNOWN
-                        && getSettings().getRequiredClusterType() != event.getNewDescription().getClusterType()) {
-                    descriptionToPublish = null;
+                        && getSettings().getRequiredClusterType() != newDescription.getClusterType()) {
+                    newDescription = null;
                 } else if (getSettings().getRequiredClusterType() == ClusterType.REPLICA_SET
                         && getSettings().getRequiredReplicaSetName() != null) {
-                    if (!getSettings().getRequiredReplicaSetName().equals(event.getNewDescription().getSetName())) {
-                        descriptionToPublish = null;
+                    if (!getSettings().getRequiredReplicaSetName().equals(newDescription.getSetName())) {
+                        newDescription = ServerDescription.builder(newDescription).type(ServerType.UNKNOWN).setName(null).build();
+                        publishDescription(ClusterType.UNKNOWN, newDescription);
```

Comment:
Does the spec require that the cluster type return to unknown?  If not, maybe we should just leave it as is.

Reply:
Not sure, failing to do so fails the test - will investigate

Reply:
`Direct connection to RSPrimary with wrong set name` directs that is set to unknown and then can be redetermined.
https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#verifying-setname-with-topologytype-single